src/sentry/services/hybrid_cloud/tombstone/impl.py

```diff
@@ -23,17 +25,32 @@ class MonolithTombstoneService(TombstoneService):
     # In the future, no single deployment can be a source of truth about the location of all models due to
     # deployment drift, however the current deployment (for which this is a bridging implementation), can.
     # We use silo limits information to infer the correct destination.
+    by_table_name: MutableMapping[str, Any]
 
-    def record_remote_tombstone(self, tombstone: ApiTombstone) -> None:
+    def __init__(self) -> None:
+        super().__init__()
+        self.by_table_name = {}
+
+    def _get_model(self, table_name: str) -> Any:
         from django.apps import apps
 
+        if model := self.by_table_name.get(table_name):
+            return model
+
         for app, app_models in apps.all_models.items():
             for model in app_models.values():
-                if limits := getattr(getattr(model, "_meta"), "silo_limits", None):
-                    if SiloMode.CONTROL in limits.silos:
-                        RegionTombstone.record_delete(tombstone.table_name, tombstone.identifier)
-                    else:
-                        ControlTombstone.record_delete(tombstone.table_name, tombstone.identifier)
+                if model._meta.db_table == table_name:
+                    if hasattr(model._meta, "silo_limit"):
+                        self.by_table_name[table_name] = model
+                        return model
+        raise ValueError(f"Could not find model by table name {table_name}")
+
+    def record_remote_tombstone(self, tombstone: ApiTombstone) -> None:
+        model = self._get_model(tombstone.table_name)
+        if SiloMode.CONTROL in model._meta.silo_limit.modes:
+            RegionTombstone.record_delete(tombstone.table_name, tombstone.identifier)
+        else:
+            ControlTombstone.record_delete(tombstone.table_name, tombstone.identifier)
```

Comment:
What do we gain from having a similar but separate local record + publish + completion from outbox messages? Having to maintain separate watermark points seems more complex than outbox messages that are removed when an operation has been propagated out.

Reply:
That doesn't quite solve the problem, because we're tracking the various ways that *relations* can be created to a tombstone.  When relations to a tombstoned model is created, there is no outbox involved necessarily.  There might be, but there might not be.

If user is deleted, both the region silo could still create new `GroupAssignee` (no outboxes involved) before or after the `User` tombstone outbox is processed (especially concurrently).   This can happen to any user relation in fact; some of them may have outboxes involved in their creation, but some may not.  This is way more likely than it seems; there are plenty of ways in which a user id can be in memory, somewhere, passed along, and recorded "as is".  We're not likely going to enforce a strict synchronous "validate user id" check on every single hc fk, that would defeat the point (it also STILL would not be strong enough, because, well, async and no spanning transactional guarantees).

So as complex as the multiple watermarks are, *it works* in all scenarios, failure and delayed.

The only other option, to say, enforce that a relation cannot be created after a tombstone is processed, would be *database level triggers* that check that an appropriate tombstone doesn't exist at the time of insert or update of a hk fc.  But that's off the table, this is our next best bet.  Also, constraints don't work here -- they cannot do joins across tables, only a trigger could.

Reply:
My point is; regardless of how we delete or process outboxes vs tombstones, the watermarks have to remain -- they track individual relation processing states.